hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +49,362 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+    ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(), access_location::host, access_mode::read);
+    ArrayHandle<int3> h_image(m_pdata->getImages(), access_location::host, access_mode::read);
+    const BoxDim& global_box = m_pdata->getGlobalBox();
+
+    ArrayHandle<Scalar3> h_linmom_accum(m_linmom_accum,
+                                        access_location::host,
+                                        access_mode::readwrite);
+    ArrayHandle<Scalar3> h_angmom_accum(m_angmom_accum,
+                                        access_location::host,
+                                        access_mode::readwrite);
+
+    for (unsigned int idx = 0; idx < num_group; ++idx)
+        {
+        // get the index from the embedded group and check if in a rigid body
+        const unsigned int particle_index = h_embed_group.data[idx];
+        const unsigned int central_tag = h_body.data[particle_index];
+        if (central_tag >= MIN_FLOPPY)
+            {
+            continue;
+            }
+        const unsigned int central_idx = h_rtag.data[central_tag];
+        // if the central particle is not local, cannot read or write to it.
+        assert(central_idx != NOT_LOCAL);
+
+        // collision on central particle itself already taken care of by collision rule
+        if (particle_index == central_idx)
+            {
+            continue;
+            }
+
+        // get velocities and masses
+        const Scalar4 vel_mass_const = h_velocity.data[particle_index];
+        const vec3<Scalar> vel_const(vel_mass_const);
+        const Scalar mass_const = vel_mass_const.w;
+
+        // get displacement
+        const Scalar4 postype_const = h_postype.data[particle_index];
+        const vec3<Scalar> pos_const(postype_const);
+        const int3 img_const = h_image.data[particle_index];
+
+        const Scalar4 postype_central = h_postype.data[central_idx];
+        const vec3<Scalar> pos_central(postype_central);
+        const int3 img_central = h_image.data[central_idx];
+
+        vec3<Scalar> displacement = pos_const - pos_central;
+        const int3 displacement_img = make_int3(img_const.x - img_central.x,
+                                                img_const.y - img_central.y,
+                                                img_const.z - img_central.z);
+        displacement = global_box.shift(displacement, displacement_img);
+
+        // change in linear and angular momentum
+        const vec3<Scalar> initial_vel_const(h_initial_vel.data[idx]);
+        const vec3<Scalar> linmom_change = mass_const * (vel_const - initial_vel_const);
+        const vec3<Scalar> angmom_change = cross(displacement, linmom_change);
+
+        // accumulate onto central particle
+        h_linmom_accum.data[central_idx] += vec_to_scalar3(linmom_change);
+        h_angmom_accum.data[central_idx] += vec_to_scalar3(angmom_change);
+        }
+    }
+
+void mpcd::CollisionMethod::transferRigidBodyMomenta(uint64_t timestep)
```

Comment:
This method may be able to use the center look-up array you found effectively if we can also figure out how to write an MPI communication method for the accumulated momentum that only operates on the center particles. We will need to look at `md::Communicator` for guidance when we get to that part. In the meantime, I think you can leave the implementation as-is on both CPU and GPU.